### DISCUSSION TITLE
How to use pyinstaller  when import from tensorflow

### DISCUSSION THREAD

**Discussion by lastrocer** [pyinstaller/pyinstaller#6890] [Help]
Can anyone provide the example to  use pyinstaller make exe file for python file witch import from Tensorflow 2.9.1
in my code
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense,LSTM,Dropout
from tensorflow.keras.optimizers import Adam

get warning message like these
40598 WARNING: Hidden import "tensorflow._api.v2.compat.v1.estimator.tpu.experimental" not found!
40613 WARNING: Hidden import "tensorflow._api.v2.compat.v2.keras.__internal__.utils" not found!
40644 WARNING: Hidden import "tensorflow._api.v2.compat.v1.keras.__internal__" not found!
40676 WARNING: Hidden import "tensorflow._api.v2.compat.v1.keras.applications.efficientnet_v2" not found!
40676 WARNING: Hidden import "tensorflow._api.v2.compat.v1.estimator.experimental" not found!
40691 WARNING: Hidden import "tensorflow._api.v2.compat.v1.keras.wrappers" not found!
40691 WARNING: Hidden import "tensorflow._api.v2.compat.v1.keras.applications" not found!
40755 WARNING: Hidden import "tensorflow._api.v2.compat.v1.keras.applications.inception_v3" not found!
40771 WARNING: Hidden import "tensorflow._api.v2.compat.v2.keras.applications.mobilenet_v3" not found!
40849 WARNING: Hidden import "tensorflow._api.v2.compat.v2.keras.applications" not found!
40927 WARNING: Hidden import "tensorflow._api.v2.compat.v1.keras.callbacks.experimental" not found!
41021 WARNING: Hidden import "tensorflow._api.v2.compat.v1.keras.metrics" not found!
41021 WARNING: Hidden import "tensorflow._api.v2.compat.v2.keras.__internal__.layers" not found!
41021 WARNING: Hidden import "tensorflow._api.v2.compat.v2.keras.premade" not found!
41068 WARNING: Hidden import "tensorflow._api.v2.compat.v1.keras.estimator" not found!
41083 WARNING: Hidden import "tensorflow._api.v2.compat.v2.keras.applications.mobilenet_v2" not found!
41114 WARNING: Hidden import "tensorflow._api.v2.compat.v1.keras.datasets.reuters" not found!
41114 WARNING: Hidden import "tensorflow._api.v2.compat.v1.estimator" not found!
41130 WARNING: Hidden import "tensorflow._api.v2.compat.v1.keras.initializers" not found!
41130 WARNING: Hidden import "tensorflow._api.v2.compat.v1.keras.regularizers" not found!
41130 WARNING: Hidden import "tensorflow._api.v2.compat.v2.keras.regularizers" not found!
41193 WARNING: Hidden import "tensorflow._api.v2.compat.v1.keras.preprocessing.text" not found!
41193 WARNING: Hidden import "tensorflow._api.v2.compat.v2.keras.models" not found!
41193 WARNING: Hidden import "tensorflow._api.v2.compat.v2.keras.wrappers" not found!
41193 WARNING: Hidden import "tensorflow._api.v2.compat.v1.keras.applications.imagenet_utils" not found!

any help would be appreciated!!

**Comment by rokm** [comment]
Those are just warnings due to internal module renaming/aliasing that `tensorflow` does, and nothing to really worry about. And if you have a problem with your executable, these are red herring - what we need is a traceback of the run-time error you are seeing.

---

If you are getting a `ModuleNotFoundError: No module named 'tensorflow'` error at run-time, try adding `--hiddenimport tensorflow` to your PyInstaller command (or add an `import tensorflow` to your program).

I did a quick test with 2.9.1 on linux, and it seems that with your test snippet, we end up not collecting `tensorflow` at all, in spite of the hook being called. Presumably because those modules are actually `keras.api._v2.keras.models`, `keras.api._v2.keras.layers`, and `keras.api._v2.keras.optimizers`, and with all the redirections, the `tensorflow` node gets no references in the modulegraph and ends up discarded...

**Reply by lastrocer** [reply]
import tensorflow not working but --hiddenimport tensorflow did the trick. Big help!!!! Thank you so much.

**Reply by rokm** [reply]
Next release of `pyinstaller-hooks-contrib` should take care of this automatically (https://github.com/pyinstaller/pyinstaller-hooks-contrib/pull/451).
